pinot-perf/src/main/java/com/linkedin/pinot/perf/BenchmarkQueryEngine.java

```diff
@@ -77,34 +110,60 @@ public void startPinot() throws Exception {
       _perfBenchmarkDriver.addSegment(segmentMetadata);
     }
 
-    System.out.println("Waiting for 10s for everything to be loaded");
-    Uninterruptibles.sleepUninterruptibly(10, TimeUnit.SECONDS);
+    ZkClient client = new ZkClient("localhost:2191", 10000, 10000, new ZNRecordSerializer());
+
+    ZNRecord record = client.readData("/PinotPerfTestCluster/EXTERNALVIEW/" + TABLE_NAME);
+    while (true) {
+      System.out.println("record = " + record);
+      Uninterruptibles.sleepUninterruptibly(10, TimeUnit.SECONDS);
+
+      int onlineSegmentCount = 0;
+      for (Map.Entry<String, Map<String, String>> stringMapEntry : record.getMapFields().entrySet()) {
+        for (Map.Entry<String, String> stringStringEntry : stringMapEntry.getValue().entrySet()) {
+          if (stringStringEntry.getValue().equals("ONLINE")) {
+            onlineSegmentCount++;
+            break;
+          }
+        }
+      }
+
+      System.out.println(onlineSegmentCount + " segments online out of " + segments.length);
+
+      if (onlineSegmentCount == segments.length) {
+        break;
+      }
+
+      record = client.readData("/PinotPerfTestCluster/EXTERNALVIEW/" + TABLE_NAME);
+    }
+
+    ranOnce = false;
+
+    System.out.println(
+        _perfBenchmarkDriver.postQuery(QUERY_PATTERNS[queryPattern], optimizationFlags).toString(2)
+    );
   }
 
   @Benchmark
   @BenchmarkMode({Mode.SampleTime})
   @OutputTimeUnit(TimeUnit.MILLISECONDS)
   public int sendQueryToPinot() throws Exception {
-    JSONObject returnValue = _perfBenchmarkDriver.postQuery(QUERY);
-    if (returnValue.getInt("numDocsScanned") != EXPECTED_NUM_DOCS_SCANNED || returnValue.getInt("totalDocs") != EXPECTED_TOTAL_DOCS_COUNT) {
-      System.out.println("returnValue = " + returnValue);
-      throw new RuntimeException("Unexpected number of docs scanned/total docs");
-    }
+    JSONObject returnValue = _perfBenchmarkDriver.postQuery(QUERY_PATTERNS[queryPattern], optimizationFlags);
     return returnValue.getInt("totalDocs");
   }
 
   public static void main(String[] args) throws Exception {
```

Comment:
If possible (I don't know how the query engine worked :-P), maybe better to pass TABLE_NAME, DATA_DIRECTORY etc from the arguments?